src/sentry/profiles/consumer.py

```diff
@@ -0,0 +1,143 @@
+            {
+                "device_os_build_number": profile["device_os_build_number"],
+                "profile": profile["sampled_profile"],
+                "symbols": profile["debug_meta"],
+            }
+        )
+
+    return normalized_profile
+
+
+def _validate_ios_profile(profile: MutableMapping[str, Any]) -> bool:
+    return "sampled_profile" in profile and "samples" in profile["sampled_profile"]
+
+
+class ProfilesWorker(AbstractBatchWorker):  # type: ignore
+    def __init__(self, producer: Producer) -> None:
+        self.__producer = producer
+        self.__producer_topic = "processed-profiles"
+
+    def process_message(self, message: Any) -> MutableMapping[str, Any]:
+        message = msgpack.unpackb(message.value(), use_list=False)
+        profile = json.loads(message["payload"])
+
+        profile.update(
+            {
+                "organization_id": message["organization_id"],
+                "project_id": message["project_id"],
+                "received": message["received"],
+            }
+        )
+
+        if profile["platform"] == "ios":
+            if not _validate_ios_profile(profile):
+                return None
+            profile = self.symbolicate(profile)
+        return _normalize(profile)
+
+    def symbolicate(self, profile: MutableMapping[str, Any]) -> MutableMapping[str, Any]:
+        samples = profile["sampled_profiles"]["samples"]
+
+        # collect all unsymbolicated frames
+        frames_by_address = {}
+        indexes_by_address = defaultdict(list)
+        for i, s in enumerate(samples):
+            for j, f in enumerate(s["frames"]):
+                indexes_by_address[f["instruction_addr"]].append((i, j))
+                frames_by_address[f["instruction_addr"]] = f
+
+        # set proper keys for process_payload to do its job
+        profile["platform"] = "native"
+        profile["stacktraces"] = {"frames": frames_by_address.values()}
+        profile["event_id"] = profile["transaction_id"]
+        profile["project"] = profile["project_id"]
+
+        # symbolicate
+        profile = process_payload(profile)
+
+        # replace  unsymbolicated frames by symbolicated ones
+        frames_by_address = {f["instruction_addr"]: f for f in profile["stacktraces"]["frames"]}
+        for address, indexes in indexes_by_address:
+            samples[indexes[0]][indexes[1]] = frames_by_address[address]
+
+        # remove unneeded keys
+        for k in ("event_id", "project", "stacktraces", "debug_meta"):
+            profile.pop(k, None)
+
+        profile["platform"] = "ios"
+
+        return profile
+
+    def flush_batch(self, profiles: Sequence[MutableMapping[str, Any]]) -> None:
+        self.__producer.poll(0)
+        for profile in profiles:
+            self.__producer.produce(
+                topic=self.__producer_topic, value=json.dumps(profile), callback=self.callback
+            )
+        self.__producer.flush()
+
+    def shutdown(self) -> None:
+        pass
```

Comment:
Not sure how important it is, but should we `.close()` the producer here?

Reply:
We ended up removing the producer from this consumer so this does not apply anymore.

Besides that, the producer class from the confluent_kafka package I was using does not have a close method, you just flush to get rid of the state, which is what I was doing right after pushing a batch of messages so no need to call it here.